airflow/traces/otel_tracer.py

```diff
@@ -0,0 +1,316 @@
+                resource=resource, id_generator=AirflowOtelIdGenerator(span_id=span_id, trace_id=trace_id)
+            )
+        else:
+            tracer_provider = TracerProvider(resource=resource)
+        tracer_provider.add_span_processor(self.span_processor)
+        tracer = tracer_provider.get_tracer(component)
+        """
+        Tracer will produce a single ID value if value is provided. Note that this is one-time only, so any
+        subsequent call will produce the normal random ids.
+        """
+        return tracer
+
+    def get_current_span(self):
+        return trace.get_current_span()
+
+    def use_span(self, span: Span):
+        return trace.use_span(span=span)
+
+    def start_span(
+        self,
+        span_name: str,
+        component: str | None = None,
+        parent_sc: SpanContext | None = None,
+        span_id=None,
+        links=None,
+        start_time=None,
+    ):
+        """Start a span; if service_name is not given, otel_service is used."""
+        if component is None:
+            component = self.otel_service
+
+        trace_id = self.get_current_span().get_span_context().trace_id
+        tracer = self.get_tracer(component=component, trace_id=trace_id, span_id=span_id)
+
+        attributes = parse_tracestate(self.tag_string) if self.tag_string else {}
+
+        if links is not None:
+            _links = gen_links_from_kv_list(links)
+        else:
+            _links = []
+
+        if start_time is not None:
+            start_time = datetime_to_nano(start_time)
+
+        if parent_sc is not None:
+            ctx = trace.set_span_in_context(NonRecordingSpan(parent_sc))
+            span = tracer.start_as_current_span(
+                span_name, context=ctx, attributes=attributes, links=_links, start_time=start_time
+            )
+        else:
+            span = tracer.start_as_current_span(
+                span_name, attributes=attributes, links=_links, start_time=start_time
+            )
+        return span
+
+    def start_span_from_dagrun(
+        self, dagrun, span_name: str | None = None, component: str = "dagrun", links=None
+    ):
+        """Produce a span from dag run."""
+        # check if dagrun has configs
+        conf = dagrun.conf
+        trace_id = int(gen_trace_id(dag_run=dagrun, as_int=True))
+        span_id = int(gen_dag_span_id(dag_run=dagrun, as_int=True))
+
+        if conf is not None:
+            traceparent = conf.get(TRACEPARENT)
+            tracestate = conf.get(TRACESTATE)
+
+        tracer = self.get_tracer(component=component, span_id=span_id, trace_id=trace_id)
+
+        if self.tag_string and tracestate:
+            tag_string = self.tag_string + "," + tracestate
+        else:
+            tag_string = self.tag_string or tracestate
+
+        if span_name is None:
+            span_name = dagrun.dag_id
+
+        if links is not None:
+            _links = gen_links_from_kv_list(links)
+        else:
+            _links = []
```

Comment:
I thought I mentioned this one already, but maybe not.

```suggestion
        _links = gen_links_from_kv_list(links) if links else []
```